course5/homework06/main.cc

```diff
@@ -0,0 +1,266 @@
+    String copyAssigned;
+    copyAssigned = copyAssigner;
+    String moveAssigned;
+    moveAssigned = std::move(moveAssigner);
+
+    copyAssigner[2] = 'z';
+
+    const int num_test_cases = 8;
+    std::array<const char *, num_test_cases> test_names = {
+        {
+            "default constructor",
+
+            "const char * construction",
+            "copy construction",
+            "move construction",
+
+            "const char * assignment",
+            "copy assignment",
+            "move assignment",
+
+            "[] assignment"
+        }
+    };
+    std::array<String, num_test_cases> inputs = {
+        {
+            defaultS,
+            sFromChar, copyConstructed, moveConstructed,
+            toBeAssigned, copyAssigned, moveAssigned,
+            copyAssigner
+        }
+    };
+    std::array<const char *, num_test_cases> expected_contents = {
+        {
+            "",
+            "word", "word", "movetest",
+            "assignment", "CopyAssignER", "MoveAssignER",
+            "CozyAssignER",
+        }
+    };
+
+    for (auto i = 0; i < num_test_cases; i++)
+    {
+        test(test_names[i], inputs[i], expected_contents[i]);
+    }
+
+    /* Tests for concatenation */
+    String concatStr = "Hello ";
+    String concatToConstChar = concatStr + "World!";
+    test("concat initialization w const char *", concatToConstChar, "Hello World!");
+
+    String concatStr2 = "World Again!";
+    String concatToString = concatStr + concatStr2;
+    test("concat initialization w String", concatToString, "Hello World Again!");
+
+    concatToString += " And again.";
+    test("+= w const char *", concatToString, "Hello World Again! And again.");
+
+    String concatStr3 = " And again twice.";
+    concatToString += concatStr3;
+    test("+= w String", concatToString, "Hello World Again! And again. And again twice.");
+
+    /* Test out of bounds error with indexing */
+    // String shortStr = "short";
+    // const char *outcome;
+    // const char *test_name;
+    //
+    // test_name = "index out of bounds";
+    // try
+    // {
+    //     shortStr[1000];
+    //     outcome = "Failed: ";
+    // }
+    // catch(...)
+    // {
+    //     outcome = "Passed: ";
+    // }
+    // std::cout << outcome << test_name << std::endl;
+    //
+    // test_name = "assignment to index out of bounds";
+    // try
+    // {
+    //     shortStr[1000] = '8';
+    //     outcome = "Failed: ";
+    // }
+    // catch(...)
+    // {
+    //     outcome = "Passed: ";
+    // }
+    // std::cout << outcome << test_name << std::endl;
+
+    /* Test assignment & construction corner cases */
+    /*
+    Assign empty string to empty string
+    Assign non-empty string to empty string
+    Assign empty string to non empty string
+    (repeat for copy/move, and construction)
+    Check the content and the length after every operation.
+    You can implement the test as one function, and array of input value and expected outputs.
```

Comment:
I must have missed the corner case you're hinting at -- all the tests I wrote just passed.